bionic/task_state.py

```diff
@@ -317,3 +329,54 @@ def _check_accessors_for_version_problems(self):
 
         for accessor in accessors_needing_saving:
             accessor.update_provenance()
+
+    # NOTE: This can be optimized further.
```

Comment:
Mostly if we can set provider as `None` once we remove provider attrs into task state. This might also benefit a lot from your tuple refactor to take stuff out from `Provider` into some sort of entity defn. Once we have a structure like that in place, we can remove usage of provider in task state completion. Also a few other stuff like setting task and queries as `None` when tasks won't be recomputed in subprocess.